### DISCUSSION TITLE
Installing apps to ~/Applications instead of /Applications

### DISCUSSION THREAD

**Discussion by milesromeno** [Homebrew/discussions#5024] [Everyday usage]
I have multiple user accounts and need applications to not be shared. I am fine with the added storage cost of duplicate installations of the same application.

Intending for each user to have their own homebrew installation, I installed homebrew using this command 
```
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
echo 'export PATH="$HOME/homebrew/bin:$PATH"' >> .zprofile
```

However, when I install applications, I can see that that they are being added to the root `/Applications` instead of my local user's `~/Applications`.
```
brew install slack
==> Downloading https://downloads.slack-edge.com/releases/macos/4.36.136/prod/arm64/Slack-4.36.136-ma
############################################################################################## 100.0%
==> Installing Cask slack
==> Moving App 'Slack.app' to '/Applications/Slack.app'
🍺  slack was successfully installed!
```
> `Moving App 'Slack.app' to '/Applications/Slack.app'`

I have found that I am able to fix this sometimes by first specifying the `HOMEBREW_CASK_OPTS` `--appdir` to be `~/Applications` and then installing
```
brew uninstall slack                                      
==> Uninstalling Cask slack
==> Backing App 'Slack.app' up to '/Users/MYUSER/homebrew/Caskroom/slack/4.36.136/Slack.app'
==> Removing App '/Applications/Slack.app'
==> Purging files for version 4.36.136 of Cask slack

export HOMEBREW_CASK_OPTS="--appdir=~/Applications"
brew install slack                               
==> Downloading https://downloads.slack-edge.com/releases/macos/4.36.136/prod/arm64/Slack-4.36.136-ma
Already downloaded: /Users/MYUSER/Library/Caches/Homebrew/downloads/683e85644f400d51e343b320335aa8065374c9d4230d5dcaee31602beff43f77--Slack-4.36.136-macOS.dmg
==> Installing Cask slack
==> Moving App 'Slack.app' to '/Users/MYUSER/Applications/Slack.app'
🍺  slack was successfully installed!
```
> `==> Moving App 'Slack.app' to '/Users/MYUSER/Applications/Slack.app'`

And I can see that the package has been added to `~/Applications`
```
ls ~/Applications 
Slack.app ...
```

However, this does not seem to work with some other applications. For example, `karabiner-elements` invariably gets installed to `/Applications` with the install output saying that it's base path is `/`
```
export HOMEBREW_CASK_OPTS="--appdir=~/Applications"
brew install karabiner-elements
==> Downloading https://github.com/pqrs-org/Karabiner-Elements/releases/download/v14.13.0/Karabiner-E
Already downloaded: /Users/MYUSER/Library/Caches/Homebrew/downloads/1713443995b53c41c8739466a95c5e109845baabbc8df2078781978e2e3fe52a--Karabiner-Elements-14.13.0.dmg
==> Installing Cask karabiner-elements
==> Running installer for karabiner-elements with sudo; the password may be necessary.
installer: Package name is Karabiner-Elements 14.13.0
installer: Installing at base path /
installer: The install was successful.
🍺  karabiner-elements was successfully installed!

ls ~/Applications 
Slack.app # karabiner-elements is not here
ls /Applications
Karabiner-Elements.app # karabiner elements is here
Karabiner-EventViewer.app
```

Why did it work as I expected with Slack, but fail with Karabiner Elements? How can I ensure that all applications are always installed to `~/Applications`?

My ultimate goal is to avoid applications installed by user A from showing up in spotlight search while user B is logged in. My current impression of how MacOS handles applications suggests the way to do that is ensure they the applications are installed in `~/Applications` rather than `/Applications`.

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /opt/homebrew.

Many of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Output of `brew config`
```
HOMEBREW_VERSION: 4.2.2-18-gdcd267b
ORIGIN: https://github.com/Homebrew/brew
HEAD: dcd267ba4caecb253220cb02a8a2bb1c46a7a65b
Last commit: 24 hours ago
Core tap JSON: 05 Jan 15:44 UTC
Core cask tap JSON: 05 Jan 15:44 UTC
HOMEBREW_PREFIX: /Users/MYUSER/homebrew
HOMEBREW_REPOSITORY: /Users/MYUSER/homebrew
HOMEBREW_CELLAR: /Users/MYUSER/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.GjmYpfeJWn/org.xquartz:0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /Users/MYUSER/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
```

### Description of issue

Sometimes GUI applications are installed, or linked (not sure about what is technically happening) to `~/Applications` while other times to `/Applications`. I need it to consistently be `~/Applications`

**Comment by Bo98** [comment]
> Why did it work as I expected with Slack, but fail with Karabiner Elements?

With Slack we download the .app and control moving it to the right place. For Karabiner Elements it is a pkg so there's often not control over individual directories.

We could perhaps add support to modify a pkg `-target` which will be a path prefix for _all_ installed files though is trickier to generalise considering some pkgs may not map to conventional user directories like ones that install to `/usr/local/bin`.

**Comment by milesromeno** [comment]
@Bo98 Is there a workaround I can do for now? Can I manually move the application from `/Applications` to `~/Applications`, or will that mess up future `brew uninstall`, `brew upgrade` commands targeting that application?

**Reply by SMillerDev** [reply]
It probably will mess with upgrades since the pkg will expect files to be in a certain location.

**Reply by ssbarnea** [reply]
That would be so awesome, maybe we can at least have an option for the non pkg based apps in between? Having to escalate privileges is such a PITA.
